README.md
# Crossbeams::LabelDesigner

The purpose of this project is to collect variable fields and configurations of labels from the clients in a visual way.

Clients can now design the entire label using the drawing tools.

By allowing our clients to indicate where on the designed label template they want the variables to be printed
as well as what font, font size etc. we can generate images to send to the printer to speed up the printing process.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'crossbeams-label_designer'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install crossbeams-label_designer

## Usage

TODO: Write usage instructions here

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/crossbeams-label_designer.


## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres roughly to [Semantic Versioning](http://semver.org/).


## [Unreleased]
### Added
### Changed
- Small changes to `Choices` selectbox handling in javascript for version 10.2.0
### Fixed

## [2.0.1] - 2023-02-22
### Fixed
- Static barcodes need to be included in the total count in the variables XML

## [2.0.0] - 2022-10-27
### Added
- Compound variables can render as barcodes.
### Changed

## [1.0.0] - 2021-05-23

Version 1.0. Major Konva upgrade and redesign of most aspects of designer.

- Changed persistence strategy to store shape descriptions, rather than full Konva JSON dump.
- Multiselect capabilities with alignment etc.
- Full-featured undo/redo facility.
- Remove overlapping constraints.
- Lines can only be horizontal or vertical.
- Variables position based on text's baseline, not top-left which makes lining-up text much easier.
- Error-correction options for QR codes.

### Changed
- Upgrade Konva library
- Refactor for version 1.0.0.
- Rake task to upgrade application's label designer javascript from gem.

## [0.2.4] - 2021-03-10
### Added
- White-on-black option for variables. If set, the variable will be printed using white text.
### Changed
- Barcode width format set in increments of 0.1 instead of 0.5.
### Fixed
- Select for compound fields was being cleared of items.

## [0.2.3] - 2019-06-28
### Changed
- Use the Choices.js library instead of Selectr.js library for decorating select boxes.

## [0.2.2] - 2019-06-11
### Changed
- Label margins introduced. All label work areas are reduced by (1mm * px/mm) in width and height. A border displays where the margin is.

## [0.2.1] - 2019-04-16
### Added
- Compound variables can be designed by building up a variable from other variables and arbitrary text.

## [0.2.0] - 2019-03-29
### Added
- The ability to create static barcode variables. These print as barcodes, but their value is fixed in the design, not dynamically applied at print time.

## [0.1.8] - 2019-01-04
### Added
- Barcode width factor 1.0.
### Changed
- Variables are displayed within groups. The config list for variables was changed to accommodate this and to allow for overriding with a different variable list.

## [0.1.7] - 2018-08-27
### Changed
- UI tries to avoid page scrolling. Toolbars remain on top while scrolling the canvas.
### Fixed
- Broken label after an undo operation when almost all shapes became selected.

## [0.1.6] - 2018-08-22
### Added
- Add the "Lato Light" font for use in text boxes only (not for variables). i.e. this font can be included in the background image, but variables only use fonts that are readily available to the print server.
### Changed
- Redesign of the toolbars:
    - Layout changed.
    - Using embedded SVG icons.
    - Fix behaviour for when to make the stroke width dropdown available.

## [0.1.5] - 2018-08-08
### Changed
- Font pixel sizes are boosted by a factor depending on the printer resolution (8 or 12 px/mm). This allows the label to print fonts at a more accurate size - closely matching the size when printed from a word processor.
### Fixed
- When an existing label is loaded and a shape uses a pixel size for a font that is no longer available, the closest pixel size is chosen and that font size is used instead.

## [0.1.4] - 2018-07-06
### Changed
- Konva 2.1.7: stage events are slightly changed. mousedown, click, mouseup, dblclick, touchstart, touchend, tap, dbltap will be triggered when clicked on empty areas too. This change guards against undefined shapes in events.

## [0.1.3] - 2018-06-29
- Page encoding is forced to UTF-8.

## [0.1.2] - 2018-04-10
### Added
- This changelog.
- EAN 128 barcode.
### Fixed
- Problem with group and shape ids getting out of sync.

## [0.1.1] - 2018-02-08
### Changed
- Upgrade to Ruby 2.5.
- Start to use git flow for releases.


lib/crossbeams/label_designer.rb
require 'json'
require 'dry/configurable'
require 'crossbeams/label_designer/version'
require 'crossbeams/label_designer/config'
require 'crossbeams/label_designer/page'
require 'crossbeams/label_designer/variable_xml'
require 'erubi'

module Crossbeams
  module LabelDesigner
    class Error < StandardError; end

    class Constants
      # See https://websemantics.uk/tools/convert-pixel-point-em-rem-percent/
      # for a tool to calculate pixel/font point conversions.
      # Basically no of px == no of pt * 1.33 (See DATA at end of file)
      # Then we need to adjust for the printer resolution - hence the constants for
      # 8PXMM and 12PXMM.
      FONT_SIZES_8PXMM = {
        16 => 6,
        19 => 7,
        22 => 8,
        24 => 9,
        27 => 10,
        30 => 11,
        32 => 12,
        35 => 13,
        38 => 14,
        40 => 15,
        43 => 16,
        48 => 18,
        54 => 20,
        59 => 22,
        64 => 24,
        70 => 26,
        75 => 28,
        86 => 32,
        96 => 36,
        107 => 40,
        118 => 44,
        128 => 48,
        144 => 54,
        160 => 60,
        176 => 66,
        192 => 72,
        213 => 80,
        235 => 88,
        256 => 96
      }.freeze

      FONT_SIZES_12PXMM = {
        24 => 6,
        28 => 7,
        32 => 8,
        36 => 9,
        40 => 10,
        44 => 11,
        48 => 12,
        52 => 13,
        56 => 14,
        60 => 15,
        64 => 16,
        72 => 18,
        80 => 20,
        88 => 22,
        96 => 24,
        104 => 26,
        112 => 28,
        128 => 32,
        144 => 36,
        160 => 40,
        176 => 44,
        192 => 48,
        216 => 54,
        240 => 60,
        264 => 66,
        288 => 72,
        320 => 80,
        352 => 88,
        384 => 96
      }.freeze

      BARCODE_TYPES = {
        'CODE_39' => 'CODE 39',
        'CODE_128' => 'CODE 128',
        'EAN_8' => 'EAN 8',
        'EAN_13' => 'EAN 13',
        'EAN_128' => 'EAN 128',
        'UPC_A' => 'UPCa',
        'UPC_E' => 'UPCe',
        'QR_CODE' => 'QR code',
        'PDF_417' => 'PDF 417',
        'CODE_93' => 'CODE 93',
        'UPC_EAN_EXTENSION' => 'UPC/EAN Extension',
        'RSS_14' => 'RSS 14',
        'RSS_EXPANDED' => 'RSS Expanded',
        'DATA_MATRIX' => 'Data Matrix'
      }.freeze
    end
  end
end
__END__
# FOR 8pxmm, multiply by 2. FOR 12pxmm, multiply by 3?
pts = [6,7,8,9,10,11,12,13,14,15,16,18,20,22,24,26,28,32,36,40,44,48,54,60,66,72,80,88,96]
pxs = pts.map {|p| (p * 1.33).ceil }
Hash[pts.zip pxs]
=> {
 6 => 8,
 7 => 10,
 8 => 11,
 9 => 12,
 10 => 14,
 11 => 15,
 12 => 16,
 13 => 18,
 14 => 19,
 15 => 20,
 16 => 22,
 18 => 24,
 20 => 27,
 22 => 30,
 24 => 32,
 26 => 35,
 28 => 38,
 32 => 43,
 36 => 48,
 40 => 54,
 44 => 59,
 48 => 64,
 54 => 72,
 60 => 80,
 66 => 88,
 72 => 96,
 80 => 107,
 88 => 118,
 96 => 128
}
[15] pry(main)> pxs8 = pts.map {|p| (p * 1.33).ceil * 2 }
=> [16, 20, 22, 24, 28, 30, 32, 36, 38, 40, 44, 48, 54, 60, 64, 70, 76, 86, 96, 108, 118, 128, 144, 160, 176, 192, 214, 236, 256]
[16] pry(main)> pxs8 = pts.map {|p| (p * 1.33 * 2).ceil }
=> [16, 19, 22, 24, 27, 30, 32, 35, 38, 40, 43, 48, 54, 59, 64, 70, 75, 86, 96, 107, 118, 128, 144, 160, 176, 192, 213, 235, 256]

[17] pry(main)> pxs12 = pts.map {|p| (p * 1.33).ceil * 3 }
=> [24, 30, 33, 36, 42, 45, 48, 54, 57, 60, 66, 72, 81, 90, 96, 105, 114, 129, 144, 162, 177, 192, 216, 240, 264, 288, 321, 354, 384]
[18] pry(main)> pxs12 = pts.map {|p| (p * 1.33 * 3).ceil }
=> [24, 28, 32, 36, 40, 44, 48, 52, 56, 60, 64, 72, 80, 88, 96, 104, 112, 128, 144, 160, 176, 192, 216, 240, 264, 288, 320, 352, 384]

FONT_SIZES_8PXMM = {
 16 => 6,
 19 => 7,
 22 => 8,
 24 => 9,
 27 => 10,
 30 => 11,
 32 => 12,
 35 => 13,
 38 => 14,
 40 => 15,
 43 => 16,
 48 => 18,
 54 => 20,
 59 => 22,
 64 => 24,
 70 => 26,
 75 => 28,
 86 => 32,
 96 => 36,
 107 => 40,
 118 => 44,
 128 => 48,
 144 => 54,
 160 => 60,
 176 => 66,
 192 => 72,
 213 => 80,
 235 => 88,
 256 => 96
}
FONT_SIZES_12PXMM = {
 24 => 6,
 28 => 7,
 32 => 8,
 36 => 9,
 40 => 10,
 44 => 11,
 48 => 12,
 52 => 13,
 56 => 14,
 60 => 15,
 64 => 16,
 72 => 18,
 80 => 20,
 88 => 22,
 96 => 24,
 104 => 26,
 112 => 28,
 128 => 32,
 144 => 36,
 160 => 40,
 176 => 44,
 192 => 48,
 216 => 54,
 240 => 60,
 264 => 66,
 288 => 72,
 320 => 80,
 352 => 88,
 384 => 96
}
